### PR TITLE
Feat/nftmarket mod in eco incentive

### PR DESCRIPTION
**Pull request by taiki1frsy**
These changes are made for implementing `ecosystem-incentive` here (https://github.com/UnUniFi/chain/pull/263).

What are added:
- Hook methods
- A method to get memo data and insert a object in keeper
- Change the args in one method to support a hook method

https://github.com/UnUniFi/chain/issues/218

### CODE REVIEW

File: x/nftmarket/keeper/utils.go [hunk 1]
```diff
@@ -8,3 +14,26 @@ func Contains(a []string, x string) bool {
 	}
 	return false
 }
+
+func (k Keeper) GetMemo(ctx sdk.Context) (string, error) {
+	txBytes := ctx.TxBytes()
```
Comment (jununifi) [inline x/nftmarket/keeper/utils.go:19]: Have you tested this behaviour via CLI?

File: x/nftmarket/keeper/nft_listing.go [hunk 2]
```diff
@@ -238,6 +231,16 @@ func (k Keeper) ListNft(ctx sdk.Context, msg *types.MsgListNft) error {
...
 	k.SaveNftListing(ctx, listing)
 
+	// get the memo data from Tx contains MsgListNft
+	// If memo data is empty, the hook method is not gonna be called.
+	txMemo, err := k.GetMemo(ctx)
+	if err != nil {
+		_ = fmt.Errorf(err.Error())
+	}
+	if len(txMemo) != 0 {
+		k.hooks.AfterNftListed(ctx, msg.NftId, txMemo)
```
Comment (jununifi) [inline x/nftmarket/keeper/nft_listing.go:241]: This should be only called when tx memo exists?

File: x/nftmarket/keeper/nft_listing.go [hunk 3]
```diff
@@ -238,6 +231,16 @@ func (k Keeper) ListNft(ctx sdk.Context, msg *types.MsgListNft) error {
 	}
 	k.SaveNftListing(ctx, listing)
 
+	// get the memo data from Tx contains MsgListNft
+	// If memo data is empty, the hook method is not gonna be called.
+	txMemo, err := k.GetMemo(ctx)
```
Comment (jununifi) [inline x/nftmarket/keeper/nft_listing.go:236]: I haven't seen any memo use case in this way, wouldn't it better to add just memo field as custom field on NftListing message?

File: x/nftmarket/keeper/nft_listing.go [hunk 4]
```diff
@@ -318,7 +321,11 @@ func (k Keeper) CancelNftListing(ctx sdk.Context, msg *types.MsgCancelNftListing
 	}
 
 	// delete listing
-	k.DeleteNftListings(ctx, listing)
+	k.DeleteNftListing(ctx, listing)
+
+	// Call AfterNftUnlistedWithoutPayment to delete NFT ID from the ecosystem-incentive KVStore
+	// since it's unlisted.
+	k.hooks.AfterNftUnlistedWithoutPayment(ctx, listing.NftId)
```
Comment (jununifi) [inline x/nftmarket/keeper/nft_listing.go:328]: Should be checking if hooks is nil or not I guess

File: x/nftmarket/keeper/nft_listing.go [hunk 5]
```diff
@@ -493,7 +500,11 @@ func (k Keeper) EndNftListing(ctx sdk.Context, msg *types.MsgEndNftListing) erro
 		if err != nil {
 			panic(err)
 		}
-		k.DeleteNftListings(ctx, listing)
+		k.DeleteNftListing(ctx, listing)
+
+		// Call AfterNftUnlistedWithoutPayment to delete NFT ID from the ecosystem-incentive KVStore
+		// since it's unlisted.
+		k.hooks.AfterNftUnlistedWithoutPayment(ctx, listing.NftId)
```
Comment (jununifi) [inline x/nftmarket/keeper/nft_listing.go:507]: Should be checking if hooks is nil or not I guess

File: x/nftmarket/keeper/nft_listing.go [hunk 6]
```diff
@@ -736,4 +748,9 @@ func (k Keeper) ProcessPaymentWithCommissionFee(ctx sdk.Context, listingOwner sd
 	} else {
 		write()
 	}
+
+	// Call AfterNftPaymentWithCommission hook method to inform the payment is successfuly
+	// executed.
+	k.hooks.AfterNftPaymentWithCommission(ctx, nftId, sdk.NewCoin(denom, fee))
```
Comment (jununifi) [inline x/nftmarket/keeper/nft_listing.go:754]: Should be checking if hooks is nil or not I guess

File: app/app.go [hunk 7]
```diff
@@ -585,6 +586,7 @@ func NewApp(
 	)
 
 	app.cdpKeeper = *cdpKeeper.SetHooks(cdptypes.NewMultiCdpHooks(app.incentiveKeeper.Hooks()))
+	app.NftmarketKeeper = *nftmarketKeeper.SetHooks(nftmarkettypes.NewMultiNftmarketHooks())
```
[the remark this answers is not in the document]
Comment (taiki1frsy) [reply app/app.go:589]: Yes, receriver hooks in `ecosystem-incentive` module.
So in this branch, there's no hooks to be called.

File: x/nftmarket/keeper/utils.go (hunk 1 above)
Comment (taiki1frsy) [reply x/nftmarket/keeper/utils.go:19]: yes.

Comment (jununifi) [reply x/nftmarket/keeper/utils.go:19]: Can you add example CLI commands when test so that other people could test with same command. It will be time taking to compose CLI commands again. (Just recommendation)

File: x/nftmarket/keeper/hooks.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+	if k.hooks != nil {
+		k.hooks.AfterNftPaymentWithCommission(ctx, nftIdentifier, fee)
+	}
+}
+
+func (k Keeper) AfterNftUnlistedWithoutPayment(ctx sdk.Context, nftIdentifier types.NftIdentifier) {
+	if k.hooks != nil {
+		k.hooks.AfterNftUnlistedWithoutPayment(ctx, nftIdentifier)
+	}
+}
```
Comment (mkXultra) [inline x/nftmarket/keeper/hooks.go:27]: plz add event for test and mainnet debug

**Review by mkXultra — CHANGES_REQUESTED** [review]

File: x/nftmarket/keeper/hooks.go (hunk 8 above)
Comment (mkXultra) [reply x/nftmarket/keeper/hooks.go:27]: Or you should unit test for hook

File: x/nftmarket/keeper/nft_listing.go [hunk 9]
```diff
@@ -90,12 +89,6 @@ func (k Keeper) SetNftListing(ctx sdk.Context, listing types.NftListing) {
 	}
 }
 
-// call this method when you want to call DeleteNftListing
-func (k Keeper) DeleteNftListings(ctx sdk.Context, listing types.NftListing) {
-	k.DeleteNftListing(ctx, listing)
-	k.UpdateListedClass(ctx, listing)
-}
-
```
Comment (mkXultra) [inline x/nftmarket/keeper/nft_listing.go:98]: Why did you delete it?

File: x/nftmarket/keeper/nft_listing.go (hunk 3 above)
Comment (mkXultra) [reply x/nftmarket/keeper/nft_listing.go:236]: plz answer it

File: x/nftmarket/keeper/nft_listing.go (hunk 2 above)
Comment (mkXultra) [reply x/nftmarket/keeper/nft_listing.go:241]: Why is this function called?

File: x/nftmarket/keeper/hooks.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+func (k Keeper) AfterNftPaymentWithCommission(ctx sdk.Context, nftIdentifier types.NftIdentifier, fee sdk.Coin) {
+	if k.hooks != nil {
+		k.hooks.AfterNftPaymentWithCommission(ctx, nftIdentifier, fee)
+	}
+}
+
+func (k Keeper) AfterNftUnlistedWithoutPayment(ctx sdk.Context, nftIdentifier types.NftIdentifier) {
```
Comment (mkXultra) [inline x/nftmarket/keeper/hooks.go:23]: what is this?

**Comment by mkXultra** [comment]
There is no confirmation that the HOOK will be called.
Please add confirmation that it is called to llistNFT,HandleFullPaymentsPeriodEndings

**Comment by taiki1frsy** [comment]
> There is no confirmation that the HOOK will be called. Please add confirmation that it is called to llistNFT,HandleFullPaymentsPeriodEndings

(since i modified keeper method for hook in nft_listing.go, please see it first)
as you know, AfterNftListed, for example, calls just `k.hooks.AfterNftListed` in types/hooks.go as a function.
So, I wrote a test to check `k.hooks.AfterNftListed` calls the actual hook methods by mocking NftmarketHooks since it's appropriate unit to be tested, right?
So, I don't think it's essential to our original purpose.

**Comment by mkXultra** [comment]
Make sure that AfterNftListed is actually called in llistNFT
Same for HandleFullPaymentsPeriodEndings

File: x/nftmarket/keeper/keeper.go [hunk 11]
```diff
@@ -5,41 +5,57 @@ import (
...
 	}
 }
 
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
+
+// SetHooks sets the nftmarket keeper hooks
+func (k *Keeper) SetHooks(hooks types.NftmarketHooks) *Keeper {
+	if k.hooks != nil {
```
Comment (mkXultra) [inline x/nftmarket/keeper/keeper.go:56]: Is that not the wrong conditional expression?

Comment (taiki1frsy) [reply x/nftmarket/keeper/keeper.go:56]: no, because it's supposed to be called only when to create app struct to avoid conflicts of hook methods registration, i guess.
this is correct.

**Review by mkXultra — APPROVED** [review]

**Comment by mkXultra** [comment]
LGTM
